### PR TITLE
Fixing wrong stack size calculation in the mono interpreter for back edges

### LINKED ISSUE #68757: Mono interpreter: Wrong stack size calculation of basic blocks
Originally, the issue was reported here: #34196 
where test case: [JIT/Regression/JitBlue/GitHub_23411/GitHub_23411.ilproj](https://github.com/dotnet/runtime/blob/2d4f2d0c8f60d5f49e39f3ddbe1824648ee2b306/src/tests/JIT/Regression/JitBlue/GitHub_23411/GitHub_23411.il) 
fails with `Invalid IL code` error message.

After further investigation - details reported here: https://github.com/dotnet/runtime/issues/34196#issuecomment-1100662244
it has been discovered that the root cause of failures is wrong stack size calculation of basic blocks in unstructured control flows with back edges.

Repro:
```
MONO_ENV_OPTIONS=--interp artifacts/obj/mono/OSX.arm64.Debug/mono/mini/mono-sgen artifacts/tests/coreclr/OSX.arm64.Debug/JIT/Regression/JitBlue/GitHub_23411/GitHub_23411/GitHub_23411.dll
```

Output:
```
GitHub_23411.IsValidName: not enough values (0 < 1) on stack at 012f
* Assertion at /Users/ivan/repos/runtime/src/mono/mono/mini/interp/transform.c:4386, condition `td->sp >= td->stack' not met
```

@BrzVlad

### PR DESCRIPTION
**Pull request by ivanpovazan**
Introduced changes are fixing the problem of wrong stack size calculation for control flows with back edges.

This has been achieved by modifying the code generation loop in `generate_code` function to:

- Skip blocks which have uninitialised stack sizes 
- Mark emitted blocks
- When the loop finishes if the code generation is not complete (there were some skipped blocks), the code generation loop is repeated
- Following passes only handle blocks skipped in previous ones (already emitted blocks are not taken into account)

This change may impact performance.

Fixes #68757.

### CODE REVIEW

File: src/mono/mono/mini/interp/transform.h [hunk 1]
```diff
@@ -126,6 +126,9 @@ struct _InterpBasicBlock {
 	// This block has special semantics and it shouldn't be optimized away
 	int eh_block : 1;
 	int dead: 1;
+
+	// Mark the block if its code has been emitted
+	gboolean already_emitted;
```
Comment (lambdageek) [inline src/mono/mono/mini/interp/transform.h:131]: nit. use 1 bit like `dead`, `eh_block` etc, above

File: src/mono/mono/mini/interp/transform.c [hunk 2]
```diff
@@ -7097,11 +7161,11 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
...
-					    *next_ip == CEE_NEWOBJ &&
-					    ((ctor_method = interp_get_method (method, read32 (next_ip + 1), image, generic_context, error))) &&
-					    is_ok (error) &&
-					    m_class_get_parent (ctor_method->klass) == mono_defaults.multicastdelegate_class &&
-					    !strcmp (ctor_method->name, ".ctor")) {
+						*next_ip == CEE_NEWOBJ &&
+						((ctor_method = interp_get_method (method, read32 (next_ip + 1), image, generic_context, error))) &&
+						is_ok (error) &&
+						m_class_get_parent (ctor_method->klass) == mono_defaults.multicastdelegate_class &&
+						!strcmp (ctor_method->name, ".ctor")) {
```
Comment (lambdageek) [inline src/mono/mono/mini/interp/transform.c:7168]: nit: unrelated whitespace change

File: src/mono/mono/mini/interp/transform.c [hunk 3]
```diff
@@ -6733,7 +6797,7 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 		}
 		case MONO_CUSTOM_PREFIX:
 			++td->ip;
-		        switch (*td->ip) {
+				switch (*td->ip) {
```
Comment (lambdageek) [inline src/mono/mono/mini/interp/transform.c:6800]: nit: whitespace

File: src/mono/mono/mini/interp/transform.c [hunk 4]
```diff
@@ -5623,7 +5687,7 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 					td->last_ins->flags |= INTERP_INST_FLAG_PROTECTED_NEWOBJ;
 				// Parameters and this pointer are popped of the stack. The return value remains
 				td->sp -= csignature->param_count + 1;
-				 // Save the arguments for the call
+				// Save the arguments for the call
```
Comment (lambdageek) [inline src/mono/mono/mini/interp/transform.c:5690]: nit: whitespace

File: src/mono/mono/mini/interp/transform.c [hunk 5]
```diff
@@ -2723,6 +2724,12 @@ interp_inline_method (TransformData *td, MonoMethod *target_method, MonoMethodHe
 			tmp_bb->il_offset = prev_ip - prev_il_code;
 			tmp_bb = tmp_bb->next_bb;
 		}
+		// After inlining, new blocks have been added as continuation to the original 'prev' block.
+		// Once the inlining is complete, we need to mark the original block as emitted as well.
+		// NOTE: There is also a potential problem with:
+		// td->offset_to_bb[td->cdd->il_offset] which will not point to the newly added block (containing inlined call),
```
Comment (lambdageek) [inline src/mono/mono/mini/interp/transform.c:2730]: I don't understand why this is a problem? we shouldn't need to look at the old call instruction again, I think.
should we zero it out in `offset_to_bb` ?

Comment (ivanpovazan) [reply src/mono/mono/mini/interp/transform.c:2730]: The reason for the comment comes from the fact that the new code gen approach combines two different ways of advancing a basic block:
1. The original approach ie `first pass` approach - where `td->offset_to_bb` maps offsets to basic blocks and uses this information to advance, by reading the map with current `in_offset`

2. The new approach ie `next pass` approach - where `td->cbb->next` is used for advancing (when skipping already emitted blocks)

After inlining, Inlined blocks are not mapped to the `td->offset_to_bb`. They exist as next blocks of the caller block. Additionally, once the inlining is completed the `td->cbb` is actually the last inlined block.  

All this seemed as something that could cause issues with the new approach. I will rethink if that comment is still valid or not.

File: src/mono/mono/mini/interp/transform.c [hunk 6]
```diff
@@ -4429,6 +4480,18 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
...
+		// Skip blocks with uninitialized stack
+		if (td->cbb->stack_height == -1) {
+			int bb_size = bb->end-bb->start;
+			g_assert (bb_size > 0);
+			td->ip += bb_size;
+
+			skipped_bbs_present = TRUE;
+			link_bblocks = FALSE;
+			emitting = FALSE;
+			continue;
```
Comment (lambdageek) [inline src/mono/mono/mini/interp/transform.c:4492]: The loop for generate_code is depending on a lot of local state in all of these boolean flags.

It would be good to do two things:
1. add a comment at the top of the loop with all the flag variables that we expect to be reset to different values.
2. Think about if we can have a state enum that is computed just from the basic block structure and the previous state, and compute it once at the beginning of the loop so that we don't have this "assign to 2 or 3 flag booleans and goto the top of the loop" pattern which is fragile, hard to debug, and depends on the execution of a bunch of code to decide what to do next.
(don't need to do it in this PR, but maybe something for the future).

File: src/mono/mono/mini/interp/transform.c [hunk 7]
```diff
@@ -7374,6 +7437,23 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 			interp_add_ins (td, MINT_NOP);
 	}
 
+	// Mark the last block with emitting information (if not already marked)
+	// and retry code generation if there was any skipped block.
+	td->cbb->already_emitted |= emitting;
```
Comment (lambdageek) [inline src/mono/mono/mini/interp/transform.c:7442]: Why is this different from earlier:

```c
if (new_bb != NULL && td->cbb != new_bb) {
			// Mark the current block with emitting information
			td->cbb->already_emitted = emitting;
```

**Review by lambdageek — COMMENTED** [review]
lgtm,  but i'm concerned that the loop in generate_code depends on a lot of incrementally-computed boolean flags.  we should at least have a comment of what flags affect the loop, and long-term we should think about having some explicit state representation (ie enum) that's computed in a separate helper function.

File: src/mono/mono/mini/interp/transform.c (hunk 7 above)
Comment (ivanpovazan) [reply src/mono/mono/mini/interp/transform.c:7442]: The `|=` on line:7442 is required for the last blocks in order to avoid the following problem:
- in the first pass the last block is marked as emitted
- in the second pass a block is skipped (`emitting` flag set to `FALSE`)
- all the adjacent blocks have been emitted and the loop exits due to `td->ip = end;`condition
- as the loop exits, the last block's `already_emitted` is overwritten with `FALSE`

The 'earlier' case in question is protected by the first condition in the loop where all current blocks are skipped which have the `already_emitted` set, so the overwriting can never happen.

As you mentioned in your previous comments the loop is really sensitive to several boolean flags which make it harder to maintain. I agree with your proposal of introducing state enums which would bring clarification for these kind of corner cases

**Review by lambdageek — APPROVED** [review]

**Comment by srxqds** [comment]
hi, any progress on this bug fix?

**Comment by ivanpovazan** [comment]
> hi, any progress on this bug fix?

@srxqds, this PR is currently on hold due to higher priority issues in my queue. 
I will get back to it as soon as possible.

Is this issue blocking your work?
